api/src/main/java/net/countercraft/movecraft/sign/AbstractMovecraftSign.java

```diff
@@ -0,0 +1,160 @@
+package net.countercraft.movecraft.sign;
+
+import net.countercraft.movecraft.craft.Craft;
+import net.countercraft.movecraft.craft.type.CraftType;
+import net.countercraft.movecraft.util.MathUtils;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
+import org.bukkit.entity.Player;
+import org.bukkit.event.block.Action;
+import org.bukkit.event.block.SignChangeEvent;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.*;
+import java.util.function.Function;
+
+// DONE: In 1.21 signs can have multiple sides! This requires us to pass the clicked side through or well the relevant lines and the set method for the clicked side => Resolved using the SignWrapper
+/*
+ * Base class for all signs
+ *
+ * A instance of a sign needs to be registered using the register function.
+ * Signs react to the following events:
+ *  - SignChangeEvent
+ *  - PlayerInteractEvent, if the clicked block is a sign
+ *  - CraftDetectEvent
+ *  - SignTranslateEvent (if the sign is a subclass of AbstractCraftSign)
+ *
+ *  Whenenver one of those events are cought by the AbstractSignListener instance, it is attempted to retrieve the relevant AbstractMovecraftSign instance.
+ *  For that, the first line of the sign's clicked side is extracted and formatting removed. If it matches the format "foo: bar", only "foo:" will be used.
+ *  With that ident, the sign is attempted to be retrieved vy tryGet(). If that returns something, the object's relevant method is called.
+ */
+public abstract class AbstractMovecraftSign {
+
+    private static final Map<String, AbstractMovecraftSign> SIGNS = Collections.synchronizedMap(new HashMap<>());
```

Comment:
Can we move this whole registry system outside of the actual `AbstractMovecraftSign` implementation, and use a singleton instead of static? we can make it accessible from the `Movecraft` instance for now. With our move towards DI/inversion of control, we should avoid adding new static singletons like this.

Reply:
Once it's merged. We already have some registry thing used in the DataTags topic, so if we add a generic registry implemtnation (i have one somewhere) we should adjust that too

Reply:
I dont think a generic registry is a good idea - its just a naming convention. I just would prefer to seperate the registration system from the registrars

Reply:
I meant a generic registry base class

Reply:
Added a small generic registry base class that can be used for multiple registries, does it fit your needs?